pkg/cli/plugins_test.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+)
+
+var _ = Describe("resolvePluginsByKey", func() {
+
+	var (
+		plugins = makePluginsForKeys(
+			"foo.example.com/v1.0",
+			"bar.example.com/v1.0",
+			"baz.example.com/v1.0",
+			"foo.kubebuilder.io/v1.0",
+			"foo.kubebuilder.io/v2.0",
+			"bar.kubebuilder.io/v1.0",
+			"bar.kubebuilder.io/v2.0",
+		)
+		resolvedPlugins []plugin.Base
+		err             error
+	)
+
+	It("should resolve keys correctly", func() {
+		By("resolving foo.example.com/v1.0")
+		resolvedPlugins, err = resolvePluginsByKey(plugins, "foo.example.com/v1.0")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"foo.example.com/v1.0"}))
+
+		By("resolving foo.example.com")
+		resolvedPlugins, err = resolvePluginsByKey(plugins, "foo.example.com")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"foo.example.com/v1.0"}))
+
+		By("resolving baz")
+		resolvedPlugins, err = resolvePluginsByKey(plugins, "baz")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"baz.example.com/v1.0"}))
+
+		By("resolving foo/v2.0")
+		resolvedPlugins, err = resolvePluginsByKey(plugins, "foo/v2.0")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"foo.kubebuilder.io/v2.0"}))
```

Comment:
The following scenario is not working: 

```
	It("should return the correct version when just the name of the plugin is informed", func() {

		var (
			plugins = makePluginsForKeys(
				"foo.example.com/v1.0",
				"foo.example.com/v2.0",
			)
			resolvedPlugins []plugin.Base
			err             error
		)


		By("resolving foo.example.com")
		resolvedPlugins, err = resolvePluginsByKey(plugins, "foo.example.com")
		Expect(err).NotTo(HaveOccurred())
		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"foo.example.com/v2.0"}))

	})
```

should we not return the upper version in this case since we are allowing pass just the name?

Reply:
This key would be ambiguous because there are multiple major versions for key `foo.example.com`. See [this case](https://github.com/kubernetes-sigs/kubebuilder/pull/1536/files#diff-228d4ad7b6960d220e6d20f73ee048c3R65).

Reply:
Agreed that it would be a RFE